core/server/worker/src/main/java/alluxio/worker/page/PagedBlockWriter.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.page;
+
+import alluxio.client.file.CacheContext;
+import alluxio.client.file.cache.CacheManager;
+import alluxio.client.file.cache.PageId;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.network.protocol.databuffer.DataBuffer;
+import alluxio.worker.block.io.BlockWriter;
+
+import io.netty.buffer.ByteBuf;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.WritableByteChannel;
+
+/**
+ * A paged implementation of BlockWriter interface.
+ */
+public class PagedBlockWriter extends BlockWriter {
+  private static final Logger LOG = LoggerFactory.getLogger(PagedBlockWriter.class);
+
+  private final CacheManager mCacheManager;
+  private final long mBlockId;
+  private final long mPageSize;
+  private long mPosition;
+
+  PagedBlockWriter(CacheManager cacheManager, long blockId, AlluxioConfiguration conf) {
+    mCacheManager = cacheManager;
+    mBlockId = blockId;
+    mPageSize = conf.getBytes(PropertyKey.USER_CLIENT_CACHE_PAGE_SIZE);
+  }
+
+  @Override
+  public long append(ByteBuffer inputBuf) throws IOException {
+    long bytesWritten = 0;
+    while (inputBuf.hasRemaining()) {
+      PageId pageId = getPageId(bytesWritten);
+      int currentPageOffset = getCurrentPageOffset();
+      int bytesLeftInPage = getBytesLeftInPage(currentPageOffset, inputBuf.remaining());
+      byte[] page = new byte[bytesLeftInPage];
+      inputBuf.get(page);
+      if (!mCacheManager.append(pageId, currentPageOffset, page, CacheContext.defaults())) {
+        throw new IOException("Append failed for block " + mBlockId);
+      }
+      bytesWritten += bytesLeftInPage;
+    }
+    mPosition += bytesWritten;
+    return bytesWritten;
+  }
+
+  @Override
+  public long append(ByteBuf buf) throws IOException {
+    long bytesWritten = 0;
+    while (buf.readableBytes() > 0) {
+      PageId pageId = getPageId(bytesWritten);
+      int currentPageOffset = getCurrentPageOffset();
+      int bytesLeftInPage = getBytesLeftInPage(currentPageOffset, buf.readableBytes());
+      byte[] page = new byte[bytesLeftInPage];
```

Comment:
Reuse the buffer?

Reply:
which buffer? do you mean `byte[] page`?  it might be more difficult, because the current `put` and `append` do not support to specify a length, they always take the length of array as the length to write to the disk.

But it's really something we could improve for both local cache and worker.   (I found we're using the same pattern in LocalCacheFileInStream)